### PR TITLE
Reference Accounts per CAIP-10

### PR DESCRIPTION
**Pull request by pedrouid**
Using CAIP-10 (Chain Agnostic Account ID Spec) for account-links

CAIP-10: https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-10.md

### CODE REVIEW

File: doctypes/account-link.md [hunk 1]
```diff
@@ -19,14 +19,14 @@ Account links publicly associate a 3ID to other public identities, such as publi
 
 **Doctype name:** `account-link`
 
-An account-link genesis record is very simple. It only contains the identifier of the account that is being linked. In most cases this is a blockchain address, such as an ethereum, or bitcoin address. In theory it could be any type of identifier that can be associated to a public key, or smart contract. The account address (account identifier) is stored in the `owners` property as one string in an array.
+An account-link genesis record is very simple. It only contains the account id (CAIP-10 complaint) of the account that is being linked. In most cases this is a blockchain address, such as an ethereum, or bitcoin address. The account id is stored in the `owners` property as one string in an array.
```
Comment (oed) [inline doctypes/account-link.md]: Maybe add a link to the CAIP-10 spec?

Comment (oed) [reply doctypes/account-link.md]: I guess after it's merged :)

File: doctypes/account-link.md [hunk 2]
```diff
@@ -19,14 +19,14 @@ Account links publicly associate a 3ID to other public identities, such as publi
...
 
 Account-link genesis records are stored in IPFS using the [`dag-cbor`](https://github.com/ipld/js-ipld-dag-cbor/) IPLD format.
 
 ```JSON
 {
   "doctype": "account-link",
-  "owners": [<account-identifier-string>]
+  "owners": [<account-id>]
```
Comment (oed) [inline doctypes/account-link.md]: change to `<caip-10-account-id>`

File: doctypes/account-link.md [hunk 3]
```diff
@@ -54,31 +54,31 @@ When resolving an account-link that is linked to a 3ID the result looks like the
...
+  "owners": ["cosmos1t2uflqwqe0fsj0shcfkrvpukewcw40yjj6hdc0@cosmos:cosmoshub-3"],
   "content": "did:3:bafyreiecedg6ipyvwdwycdjiakhj5hiuuutxlvywtkvckwvsnu6pjbwxae"
 }
 ```
 
 ### Bitcoin account-link
 ```JSON
 {
-  "owners": ["128Lkh3S7CkDTBZ8W7BbpsN3YYizJMp8p6"],
+  "owners": ["128Lkh3S7CkDTBZ8W7BbpsN3YYizJMp8p6@bip122:000000000019d6689c085ae165831e93"],
```
Comment (oed) [inline doctypes/account-link.md]: Cool so this could be used to link to any bitcoin fork?

File: doctypes/account-link.md (hunk 1 above)
Comment (oed) [reply doctypes/account-link.md]: Looks like you beat me to it :P

File: doctypes/account-link.md (hunk 3 above)
Comment (pedrouid) [reply doctypes/account-link.md]: Yes, literally any BIP122 compatible similar to how Ethereum forks that are compatible with EIP155.

Examples from [CAIP-2](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md)
```
# Bitcoin mainnet (see https://github.com/bitcoin/bips/blob/master/bip-0122.mediawiki#definition-of-chain-id)
bip122:000000000019d6689c085ae165831e93

# Litecoin
bip122:12a765e31ffd4059bada1e25190f6e98

# Feathercoin (Litecoin fork)
bip122:fdbe99b90c90bae7505796461471d89a
```

**Comment by pedrouid** [comment]
The only thing that is wrong with this PR that should be fixed before merging, is that there isn't any specifications as far as I know to reference polkadot blockchain IDs

CAIP-2 is a very generic for blockchain IDs which in turn have more specific CAIPs that define for each blockchain namespace. For Ethereum we got [CAIP-3](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-3.md), for Bitcoin we got [CAIP-4](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-4.md)

While it's not required to create a CAIP to add Polkadot, it would be important define a more specific namespace for Polkadot account IDs other than my draft `polkadot:polkadot-chain`

**Comment by joshua-mir** [comment]
Polkadot addresses already contain a network identifier (some details here https://github.com/paritytech/substrate/wiki/External-Address-Format-(SS58)) but this is only enforced on the frontend (which converts addresses to a raw pubkey in most cases). 

other than that, there aren't any standards I'm aware of to disambiguate between chains, I think the right place to start would be to grab either [spec_name](https://github.com/paritytech/polkadot/blob/b5cc86af8cc0005c9c902156a43f688de549d3f9/runtime/kusama/src/lib.rs#L78) as declared by individual chains (which you can grab using the `system.lastRuntimeUpgrade()` storage value), or their genesis hashes.

**Comment by oed** [comment]
That's very helpful @joshua-mir! Looking at [your link](https://github.com/paritytech/substrate/wiki/External-Address-Format-(SS58)) it seems like there are only 48 address types that refer to networks. Is that right, or are you imagining that future networks when those 48 are filled up would use 128-...? 

@pedrouid do you think `polkadot:genesis-hash` or `polkadot:spec_name` makes the most sense?

@joshua-mir what would happen if two different chains use the same `spec_name`? Is that possible?

**Comment by joshua-mir** [comment]
@oed it's completely possible for two networks to share the same spec_name, and I don't believe it would cause significant issues, networking and consensus all use the genesis hash, spec_name is only used in the client (to determine which folder to look for chaindata in) and "onchain".

**Comment by shawntabrizi** [comment]
I don't think the external address format is a good identifier for a chain, since many different chains may also share them.

Something like the spec_name + genesis hash makes more sense, but at that point, genesis hash should be good enough without the spec_name I think...

**Comment by pedrouid** [comment]
So basically you dont have the concept of the chainId in Polkadot? @joshua-mir @shawntabrizi 

Then I guess this would be similar to Bitcoin's BIP122. Would it be possible to write a similar spec to CAIP-4 where we describe the Polkadot namespace for chainIds using the genesis hash?

https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-4.md

I can help with the PR but lets get it done on the CAIPs so that we can use it as blockchain reference for these account-links

**Comment by pedrouid** [comment]
I've updated the polkadot example to be CAIP-13 compatible @oed 

Let's just wait for the CAIP-13 to be merged before merging this PR

**Comment by pedrouid** [comment]
Ready to Merge!

**Pull request merged**